File: backend/src/schemas/catalog.py
"""
Catalog Schemas

Pydantic models for flat-list catalog API (Netflix-style).
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from uuid import UUID

from src.schemas.common import PaginatedResponse


class CatalogItemResponse(BaseModel):
    """
    Single catalog item (video file with context).

    This is the flat-list view - each video file is an independent item
    with its project/event context embedded.
    """
    # Video file info
    id: UUID
    display_title: Optional[str] = Field(None, description="Human-readable title")
    file_name: str = Field(..., description="Original filename")
    file_path: str = Field(..., description="Full NAS path")

    # Technical metadata
    duration_seconds: Optional[int] = None
    file_size_bytes: Optional[int] = None
    file_format: Optional[str] = None
    resolution: Optional[str] = None
    version_type: Optional[str] = None

    # Context (from joined tables)
    project_code: Optional[str] = Field(None, description="Project code (WSOP, HCL, etc.)")
    project_name: Optional[str] = Field(None, description="Project display name")
    year: Optional[int] = Field(None, description="Year from season or event")
    event_name: Optional[str] = Field(None, description="Event name")
    episode_title: Optional[str] = Field(None, description="Episode title if available")

    # Status
    is_hidden: bool = False
    hidden_reason: Optional[str] = None
    scan_status: str = "pending"

    # Timestamps
    created_at: datetime
    updated_at: datetime
    file_mtime: Optional[datetime] = None

    model_config = {"from_attributes": True}


class CatalogListResponse(PaginatedResponse[CatalogItemResponse]):
    """Paginated catalog response"""
    pass


class CatalogStatsResponse(BaseModel):
    """Catalog statistics summary"""
    total_files: int = Field(..., description="Total video files")
    visible_files: int = Field(..., description="Files visible in catalog (not hidden)")
    hidden_files: int = Field(..., description="Hidden files count")

    by_project: dict = Field(default_factory=dict, description="File count by project")
    by_year: dict = Field(default_factory=dict, description="File count by year")
    by_format: dict = Field(default_factory=dict, description="File count by format")

    total_duration_hours: Optional[float] = Field(None, description="Total duration in hours")
    total_size_gb: Optional[float] = Field(None, description="Total size in GB")


class CatalogFilterParams(BaseModel):
    """Filter parameters for catalog queries"""
    project_code: Optional[str] = Field(None, description="Filter by project code")
    year: Optional[int] = Field(None, description="Filter by year")
    search: Optional[str] = Field(None, description="Search in display_title and file_name")
    include_hidden: bool = Field(False, description="Include hidden files")
    version_type: Optional[str] = Field(None, description="Filter by version type")
    file_format: Optional[str] = Field(None, description="Filter by file format (mp4, mov, etc.)")

File: frontend/src/types/catalog.ts
/**
 * Catalog Types - 카탈로그 관련 타입 정의
 * BLOCK_FRONTEND / FrontendAgent
 */

// 카탈로그 아이템 (비디오 파일)
export interface CatalogItem {
  id: string;
  display_title: string | null;
  file_name: string;
  file_path: string;

  // 기술 메타데이터
  duration_seconds: number | null;
  file_size_bytes: number | null;
  file_format: string | null;
  resolution: string | null;
  version_type: string | null;

  // 컨텍스트
  project_code: string | null;
  project_name: string | null;
  year: number | null;
  event_name: string | null;
  episode_title: string | null;

  // 상태
  is_hidden: boolean;
  hidden_reason: string | null;
  scan_status: string;

  // 타임스탬프
  created_at: string;
  updated_at: string;
  file_mtime: string | null;
}

// 카탈로그 목록 응답
export interface CatalogListResponse {
  items: CatalogItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// 카탈로그 통계
export interface CatalogStats {
  total_files: number;
  visible_files: number;
  hidden_files: number;
  by_project: Record<string, number>;
  by_year: Record<string, number>;
  by_format: Record<string, number>;
  total_duration_hours: number | null;
  total_size_gb: number | null;
}

// 필터 옵션
export interface CatalogFilterOptions {
  projects: Array<{ code: string; name: string; count: number }>;
  years: number[];
  formats: string[];
  version_types: string[];
}

// 필터 파라미터
export interface CatalogFilterParams {
  page?: number;
  page_size?: number;
  project_code?: string;
  year?: number;
  search?: string;
  include_hidden?: boolean;
  version_type?: string;
  file_format?: string;
}

// 프로젝트 코드별 색상
export const PROJECT_COLORS: Record<string, string> = {
  WSOP: 'bg-yellow-500',
  HCL: 'bg-red-500',
  GGMILLIONS: 'bg-blue-500',
  GOG: 'bg-purple-500',
  PAD: 'bg-green-500',
  MPP: 'bg-orange-500',
  OTHER: 'bg-gray-500',
};

// 포맷별 아이콘/스타일
export const FORMAT_LABELS: Record<string, string> = {
  mp4: 'MP4',
  mov: 'MOV',
  mxf: 'MXF',
  avi: 'AVI',
  mkv: 'MKV',
};
